### PR TITLE
 [ALLUXIO-2907] Improve Alluxio UFS Test

### PR DESCRIPTION
**Pull request by apc999**
https://alluxio.atlassian.net/browse/ALLUXIO-2907

By default the UFS contract tests are not run. 
One can run a UFS (e.g. s3a) test by

```bash
$  mvn test -pl underfs/s3a -DtestS3Bucket=s3a://apc999 \
-Daws.accessKeyId=<KEY> -Daws.secretKey=<KEY>
```
or for HDFS
```bash
$ mvn test -pl underfs/hdfs -DtestHdfsBaseDir=hdfs://localhost:9000/tmp
```

### CODE REVIEW

File: underfs/s3a/src/test/java/alluxio/underfs/s3a/S3AUnderFileSystemContractTest.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    return new S3AUnderFileSystemFactory().create(path, conf);
+  }
+
+  @Override
+  public String getUfsBaseDir() {
+    mS3Bucket = PathUtils.concatPath(System.getProperty(INTEGRATION_S3_BUCKET), UUID.randomUUID());
+    Preconditions.checkState(mS3Bucket != null && !mS3Bucket.equals(""),
+        PreconditionMessage.S3_BUCKET_MUST_BE_SET.toString(), INTEGRATION_S3_BUCKET);
+    return mS3Bucket;
+  }
```
Comment (calvinjia) [inline underfs/s3a/src/test/java/alluxio/underfs/s3a/S3AUnderFileSystemContractTest.java]: Do we need a clean up as well?

File: core/common/src/test/java/alluxio/underfs/AbstractUnderFileSystemContractTest.java [hunk 2]
```diff
@@ -33,134 +23,136 @@
...
   public void createOpenEmpty() throws IOException {
-    String testFile = PathUtils.concatPath(mUnderfsAddress, "testFile");
+    String testFile = PathUtils.concatPath(mUnderfsAddress, "createOpenEmpty");
     createEmptyFile(testFile);
     byte[] buf = new byte[0];
     int bytesRead = mUfs.open(testFile).read(buf);
     // TODO(adit): Consider making the return value uniform across UFSs
     if (UnderFileSystemUtils.isHdfs(mUfs)) {
-      Assert.assertTrue(bytesRead == -1);
+      Assert.assertEquals(-1, bytesRead);
```
Comment (calvinjia) [inline core/common/src/test/java/alluxio/underfs/AbstractUnderFileSystemContractTest.java]: This should be an explicit property of the ufs.

Comment (apc999) [inline core/common/src/test/java/alluxio/underfs/AbstractUnderFileSystemContractTest.java]: the TODO above seems to be related to this

File: underfs/s3a/src/test/java/alluxio/underfs/s3a/S3AUnderFileSystemContractTest.java (hunk 1 above)
Comment (apc999) [inline underfs/s3a/src/test/java/alluxio/underfs/s3a/S3AUnderFileSystemContractTest.java]: cleanup is in `after()` in the abstract class

File: core/common/src/test/java/alluxio/underfs/AbstractUnderFileSystemContractTest.java [hunk 3]
```diff
@@ -33,134 +23,137 @@
...
+    mUnderfsAddress = PathUtils.concatPath(getUfsBaseDir(), UUID.randomUUID());
+    mUfs = createUfs(mUnderfsAddress, UnderFileSystemConfiguration.defaults());
+    mUfs.mkdirs(mUnderfsAddress, MkdirsOptions.defaults().setCreateParent(true));
   }
 
   @After
   public final void after() throws Exception {
-    ConfigurationTestUtils.resetConfiguration();
+    mUfs.deleteDirectory(mUnderfsAddress, DeleteOptions.defaults().setRecursive(true));
+    mUfs.close();
```
Comment (calvinjia) [inline core/common/src/test/java/alluxio/underfs/AbstractUnderFileSystemContractTest.java]: Will this slow down the tests? Also, does this work with HDFS's client caching?

**Comment by apc999** [comment]
@calvinjia  PTAL

File: core/common/src/test/java/alluxio/underfs/AbstractUnderFileSystemContractTest.java [hunk 4]
```diff
@@ -33,134 +23,137 @@
...
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
 import java.util.Arrays;
-import java.util.List;
+import java.util.UUID;
+
+/**
+ * This is the base class of UFS contract tests. It describes the contract of Alluxio with the UFS
+ * through the UFS interface. Each implementation of {@link UnderFileSystem} is expected to extend
```
Comment (apc999) [inline core/common/src/test/java/alluxio/underfs/AbstractUnderFileSystemContractTest.java]: done

**Comment by apc999** [comment]
I will have a follow-up PR to clean tests module up , and update the doc

**Pull request merged**